src/coreclr/jit/llvm.cpp

```diff
@@ -1654,14 +1630,190 @@ void startImportingNode(llvm::IRBuilder<>& builder)
+
+    GenTreeIntCon* shadowStackOffset = _compiler->gtNewOneConNode(var_types::TYP_UINT)->AsIntCon(); // LLVM-TODO: TYP_LONG for Wasm64?
+
+    _compiler->eeGetMethodSig(_compiler->info.compMethodHnd, &_sigInfo);
+    _sigInfoIsValid = true;
+
+    shadowStackOffset->SetIconValue(getTotalParameterOffset(_sigInfo));
+
+    // This is a placeholder that we'll replace at LLVM generation as there is no IR Node or Arg
+    GenTreeIntCon* shadowStackArg = _compiler->gtNewOneConNode(var_types::TYP_UINT)->AsIntCon();
+
+    // TODO-LLVM : is GT_ADD the right way to do pointer arithmetic?
+    GenTree* localShadowStack = _compiler->gtNewOperNode(genTreeOps::GT_ADD, var_types::TYP_REF, shadowStackArg, shadowStackOffset);
+
+    GenTreeLclVar* shadowStack = _compiler->gtNewStoreLclVar(shadowStackLclNum, localShadowStack);
+    LclVarDsc* shadowStackVarDsc = _compiler->lvaGetDesc(shadowStack->GetLclNum());
+    shadowStackVarDsc->lvIsParam = 1;
+    shadowStackVarDsc->lvType = var_types::TYP_REF;
+
+    // insert the local shadowstack offest at the beginning of the first block
+    // This position is well known and replaced at LLVM generation
+    // LLVM-TODO: This is maybe not the best strategy, how else could we identity these IR nodes - a new block?
+    BasicBlock* firstBlock = _compiler->fgFirstBB;
+    LIR::Range& firstRange = LIR::AsRange(firstBlock);
+    firstRange.InsertAtBeginning(shadowStackOffset);
+    firstRange.InsertAfter(shadowStackOffset, shadowStackArg);
+    firstRange.InsertAfter(shadowStackArg, localShadowStack);
+    firstRange.InsertAfter(localShadowStack, shadowStack);
+
+    for (BasicBlock* const block : _compiler->Blocks())
+    {
+        LIR::Range& lirRange = LIR::AsRange(block);
+        for (GenTree* node : LIR::AsRange(block))
+        {
+            genTreeOps oper = node->OperGet();
+            if (oper == GT_STORE_LCL_VAR)
+            {
+                GenTreeLclVarCommon* lclVar = node->AsLclVarCommon();
+                LclVarDsc* varDsc = _compiler->lvaGetDesc(lclVar->GetLclNum());
+                if (!varDsc->lvIsParam && !canStoreLocalOnLlvmStack(varDsc))
+                {
+                    GenTreeIntCon* offset;
+                    GenTreeLclVar* shadowStackVar;
+                    GenTree* addNode = createAddNodeForShadowStackLocal(varDsc, shadowStackLclNum, offset, shadowStackVar);
+
+                    genTreeOps oper = lclVar->TypeGet() == var_types::TYP_STRUCT ? GT_STORE_OBJ : GT_STOREIND;
+
+                    node->ChangeOper(oper);
+                    node->gtFlags |= GTF_IND_TGT_NOT_HEAP;
+                    GenTreeOp* opNode = node->AsOp();
+                    opNode->gtOp2 = opNode->gtGetOp1();
+                    opNode->gtOp1 = addNode;
+
+                    insertAddWithOperands(lirRange, node, addNode, offset, shadowStackVar);
+                }
+            }
+            else if (oper == GT_LCL_VAR)
```

Comment:
This code looks a bit repetitive, I wonder if we could deduplicate it with something like this:
```cpp
if (!node->OperIs(GT_LCL_VAR, GT_STORE_VAR))
{
    continue;
}

GenTreeLvlVarCommon* lclVar = ...;
// The varDsc, the createAddress call, etc

genTreeOps indirOper = GT_NONE:
switch (node->OperGet() { /* choose the new oper */ }

node->ChangeOper(indirOper);
node->AsIndir()->SetAddr(address);

if (GenTree::OperIsStore(indiOper)
{
   node->AsOp()->gtOp2 = storedValue;
}
```